### PR TITLE
Enable pretty-printers to print stake amounts in tokens

### LINKED ISSUE #3037: Display stake amount in token units in CLI command output
##### SUMMARY

To have a better UX when using the `oasis-node` CLI with a hardware wallet (e.g. Ledger) or on its own, it should also support displaying stake amounts in token unit (rather than base units used internally).

For example, `oasis stake info` currently displays something like:
```
Total supply: 10000000000000000000
Common pool: 7999217230119682890
Last block fees: 0
Staking threshold (entity): 100000000000
Staking threshold (node-validator): 100000000000
Staking threshold (node-compute): 100000000000
Staking threshold (node-storage): 100000000000
Staking threshold (node-keymanager): 100000000000
Staking threshold (runtime-compute): 100000000000
Staking threshold (runtime-keymanager): 100000000000
```

It could display something like:
```
Total supply: AMBER 10000000000
Common pool: AMBER 7999217230.11968289
Last block fees: 0
Staking threshold (entity): AMBER 100
Staking threshold (node-validator): AMBER 100
Staking threshold (node-compute): AMBER 100
Staking threshold (node-storage): AMBER 100
Staking threshold (node-keymanager): AMBER 100
Staking threshold (runtime-compute): AMBER 100
Staking threshold (runtime-keymanager): AMBER 100
```

where `AMBER` is the token's ticker symbol for the currently running [Oasis Amber Network](https://docs.oasis.dev/operators/amber-network.html).

We could support configuring both:
- amount of *base token* units one token unit is worth (e.g. 1000000000),
- token's ticker symbol (e.g. `AMBER`).

##### ISSUE TYPE
- Feature Idea

### PR DESCRIPTION
**Pull request by tjanez**
Closes #3037.

Example outputs of update `oasis-node stake` CLI commands:

```
[tadej@toronto staking-docs-network]$ oasis-node-dev stake info -a $ADDR
Token's ticker symbol: AMBER
Token's value base-10 exponent: 9
Total supply: AMBER 10000000000.0
Common pool: AMBER 7999217230.11968289
Last block fees: AMBER 0.0
Staking threshold (entity): AMBER 100.0
Staking threshold (node-validator): AMBER 100.0
Staking threshold (node-compute): AMBER 100.0
Staking threshold (node-storage): AMBER 100.0
Staking threshold (node-keymanager): AMBER 100.0
Staking threshold (runtime-compute): AMBER 100.0
Staking threshold (runtime-keymanager): AMBER 100.0
```

```
[tadej@toronto staking-docs-network]$ oasis-node-dev stake account info   -a $ADDR   --stake.account.address oasis1qp26ua0g6t30eus5mkxuv08fkds37jcqs5fs9tu7
General Account:
  Balance: AMBER 10000.000000001
  Nonce: 0
Escrow Account:
  Active:
    Balance: AMBER 0.0
    Total Shares: 0
  Debonding:
    Balance: AMBER 0.0
    Total Shares: 0
  Commission Schedule:
    Rates: (none)
    Rate Bounds: (none)
  Stake Accumulator:
    Claims: (none)
```

### CODE REVIEW

File: go/oasis-node/cmd/stake/stake.go [hunk 1]
```diff
@@ -90,6 +90,28 @@ func doWithRetries(cmd *cobra.Command, descr string, fn func() error) {
 	os.Exit(1)
 }
 
+func getTokenSymbol(ctx context.Context, cmd *cobra.Command, client api.Backend) string {
+	var symbol string
+	doWithRetries(cmd, "query token's symbol", func() error {
```
Comment (kostko) [inline go/oasis-node/cmd/stake/stake.go:95]: What's up with these retries all over the staking cmds, these should never be needed. Can we open an issue to get rid of this and not introduce more in new code?

File: go/oasis-test-runner/scenario/e2e/stake_cli.go [hunk 2]
```diff
@@ -81,6 +82,8 @@ var (
 	// we can expect the amount of reclaimed base units to equal the number of reclaimed escrow
 	// shares.
 	reclaimEscrowAmount int64 = reclaimEscrowShares
+
+	ctx = context.Background()
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/stake_cli.go:86]: Having a global like this sounds bad as `ctx` is commonly used and this makes it global in the whole `e2e` package.

File: go/oasis-test-runner/scenario/e2e/stake_cli.go [hunk 3]
```diff
@@ -826,3 +828,9 @@ func (sc *stakeCLIImpl) genAmendCommissionScheduleTx(childEnv *env.Env, nonce in
 	}
 	return nil
 }
+
+func init() {
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/stake_cli.go:832]: Why do this in global init, I would much rather have this per test case.

File: go/staking/api/api.go [hunk 4]
```diff
@@ -88,6 +94,13 @@ var (
 		MethodAmendCommissionSchedule,
 	}
 
+	// ContextKeyTokenSymbol is the key to retrieve the token's ticker symbol
+	// value from a context.
+	ContextKeyTokenSymbol = contextKey("staking/token-symbol")
```
Comment (kostko) [inline go/staking/api/api.go:99]: The `Context` prefix sounds too general as we have other kinds of contexts. Maybe explicitly name this `PrettyPrinterContext...`.

Also consider moving all this pretty printing stuff into its own `prettyprint.go` file.

File: go/oasis-node/cmd/stake/stake.go (hunk 1 above)
Comment (tjanez) [reply go/oasis-node/cmd/stake/stake.go]: Old code fixed in #3113. I've also removed references in this PR.

File: go/oasis-test-runner/scenario/e2e/stake_cli.go (hunk 2 above)
Comment (tjanez) [reply go/oasis-test-runner/scenario/e2e/stake_cli.go]: Agreed. I've made it per test case now.

File: go/oasis-test-runner/scenario/e2e/stake_cli.go (hunk 3 above)
Comment (tjanez) [reply go/oasis-test-runner/scenario/e2e/stake_cli.go]: Agreed. I've made it per test case now.

File: go/staking/api/api.go (hunk 4 above)
Comment (tjanez) [reply go/staking/api/api.go]: > The `Context` prefix sounds too general as we have other kinds of contexts. Maybe explicitly name this `PrettyPrinterContext...`.

Yes, good point. I've renamed it.

> Also consider moving all this pretty printing stuff into its own `prettyprint.go` file.

Did you mean just the new `ConvertToTokenAmount()` and `PrettyPrintAmount()` functions?

For `PrettyPrint()` and `PrettyType()` methods, I would rather have them defined in the same place as the corresponding types.

If we would like to make the `go/staking/api/api.go` file smaller, I would rather break out logical pieces (e.g. events-related stuff, transaction-related stuff, threshold-related stuff, or similar).

Comment (kostko) [reply go/staking/api/api.go]: I mean just the new functions.

Comment (tjanez) [reply go/staking/api/api.go]: > I mean just the new functions (and context keys).

Aha, ok. I can move those to `prettyprint.go`.

Comment (tjanez) [reply go/staking/api/api.go]: > > I mean just the new functions (and context keys).
> Aha, ok. I can move those to `prettyprint.go`.

Done.

**Pull request merged**